src/Gemstone.IO/Collections/FileBackedLookupTable.cs

```diff
@@ -1748,45 +1730,143 @@ static FileBackedLookupTable()
 
         private static Func<Stream, T>? GetReadMethod<T>()
         {
             Type type = typeof(T);
             ConstructorInfo? constructor = type.GetConstructor(Flags, null, s_types, null);
 
-            if (constructor is null)
+            if (constructor is not null)
             {
-                constructor = type.GetConstructor(Flags, null, Type.EmptyTypes, null);
+                List<ParameterExpression> parameterExpressions = s_types.Select(Expression.Parameter).ToList();
+                NewExpression newExpression = Expression.New(constructor, parameterExpressions);
+                LambdaExpression lambdaExpression = Expression.Lambda(typeof(Func<Stream, T>), newExpression, parameterExpressions);
+                return (Func<Stream, T>)lambdaExpression.Compile();
+            }
 
-                if (constructor is null)
-                    return null;
+            constructor = type.GetConstructor(Flags, null, Type.EmptyTypes, null);
 
+            if (constructor is not null)
+            {
                 MethodInfo? method = type.GetMethod("ReadFrom", Flags, null, s_types, null);
 
-                if (method is null)
-                    return null;
+                if (method is not null)
+                {
+                    return stream =>
+                    {
+                        T obj = Activator.CreateInstance<T>();
+                        method.Invoke(obj, new object[] { stream });
+                        return obj;
+                    };
+                }
+            }
+
+            Func<BinaryReader, T>? binaryReaderFunc = GetBinaryReaderMethod<T>();
 
+            if (binaryReaderFunc is not null)
+            {
                 return stream =>
                 {
-                    T obj = Activator.CreateInstance<T>();
-                    method.Invoke(obj, new object[] { stream });
-
-                    return obj;
+                    using BinaryReader reader = new BinaryReader(stream, Encoding.UTF8, true);
+                    return binaryReaderFunc(reader);
                 };
             }
 
-            List<ParameterExpression> parameterExpressions = s_types.Select(Expression.Parameter).ToList();
-            NewExpression newExpression = Expression.New(constructor, parameterExpressions);
-            LambdaExpression lambdaExpression = Expression.Lambda(typeof(Func<Stream, T>), newExpression, parameterExpressions);
+            return null;
+        }
 
-            return (Func<Stream, T>)lambdaExpression.Compile();
+        private static Action<BinaryWriter, T>? GetBinaryWriterMethod<T>()
+        {
+            void WriteDateTime(BinaryWriter writer, DateTime dt)
+            {
+                writer.Write((byte)dt.Kind);
+                writer.Write(dt.Ticks);
+            }
+
+            void WriteString(BinaryWriter writer, string? str)
+            {
+                writer.Write(str ?? string.Empty);
+
+                if (string.IsNullOrEmpty(str))
```

Comment:
Shouldn't this be `str is null` instead of `string.IsNullOrEmpty(str)` - otherwise `ReadString` will return `null` for any empty string?

Reply:
It needs to be `string.IsNullOrEmpty()`. The idea to serialize empty string for both null and empty string, then encode a byte that tells us whether it's null or empty string.

Maybe it would make more sense like this?

```c#
void WriteString(BinaryWriter writer, string? str)
{
    if (string.IsNullOrEmpty(str))
    {
        writer.Write(str ?? string.Empty);
        writer.Write(str is null);
        return;
    }

    writer.Write(str);
}
```